### ISSUE TITLE
Offline access

### ISSUE THREAD

**Issue by BSFishy** [NoahBres/LearnRoadRunner#15]
The documentation has been a great help in getting RoadRunner up an running quickly, but the most annoying part is having to constantly switch between the robot's wifi and the internet so that I can look at the next section. It's not a super huge issue, but it is just a bit aggravating having to switch to a network that has internet access to see the next section and then switch back to the robot's wifi so I can change the constants and whatnot.

I'm not super sure if it is possible or how possible it is, but I would recommend somehow making the page a progressive web app or single page application.

I think having a PWA would probably be easier because all you really need, I think, is a `manifest.json` file that explains the files that need to be downloaded for offline access. Then, it would also be able to be downloaded as if it were a regular program (although I believe this can be disabled), but most importantly it would enable the entire site to be accessed while offline, so it wouldn't matter if one is on the robot wifi or connected to the internet.

Using a SPA would work as well, but I'm not sure how well it would fit into your current stack. Essentially, from what I know, the page is loaded initially with all of the pages it needs and just displays those pages when needed. So, instead of using multiple pages, it is essentially just one page that has everything it needs in memory. This way, once you load the page initially, you can access the entire website so long as it is in memory.

I would be willing to assist in bringing this into fruition, but again I'm not super duper familiar with the stack.

**Comment by BSFishy** [comment]
After looking a bit into it, it looks like it should be super easy to make it a PWA and work offline with [`@vuepress/plugin-pwa`](https://vuepress.vuejs.org/plugin/official/plugin-pwa.html#install).

**Comment by NoahBres** [comment]
This is a really neat idea! The vuepress pwa plugin actually looks very seamless! The @vuepress/plugin-pwa page seems to just imply that all you need to do is set install, set serviceWorker to true, and it'll simply start caching everything with a service worker? The only issue is that it seems to only cache rather than download ahead of time. Not entirely sure though. Would have to look into that. If that were the case, you'd have to visit every page to pre-populate the cache before going offline.

I will most definitely look into it when I have the time! 

For the time being, you are able to get offline access if you clone and just run it locally. Since you're familiar with web dev.

**Comment by BSFishy** [comment]
Just curious, is there any update on this? I see there was a commit a couple days ago that added the plugin and stuff, but I don't see anything on the production site. I also see some changes that were commit and are on the production site. Is something maybe not working, or is it just not enabled or something? Again, I'm just curious about the state of this.

**Comment by NoahBres** [comment]
Honestly, no clue.
I assumed that the PWA plugin would enable offline cache via the service workers automatically with zero config but it doesn't seem to actually do anything?

**Comment by NoahBres** [comment]
Might just have to try and work on implementing offline caching via service workers from scratch.

**Comment by BSFishy** [comment]
From the `Application` tab of Inspect Elements, it gives the following errors:
 * `Manifest start URL is not valid`
 * `No matching service worker detected. You may need to reload the page, or check that the scope of the service worker for the current page encloses the scope and start URL from the manifest.`
 * `Downloaded icon was empty or corrupted`

In addition to an error in the log:
 * `Error while trying to use the following icon from the Manifest: https://www.learnroadrunner.com/android-chrome-192x192.png (Download error or resource isn't a valid image)`

To me, it looks like there is no `start_url` property in the `manifest.json` file as well as the icons in there don't have the `/assets` at the beginning. There might also be an issue with the service worker, but that may be an issue in the build process, which I haven't checked yet.

**Comment by BSFishy** [comment]
After pulling and making some small changes, it seems that making those 2 changes to the `manifest.json` file will fix 2 of the issues with installing. The last seems to be with the build process. According to this line:

```javascript
  if (process.env.NODE_ENV === 'production' && !isServer && SW_ENABLED) {
```

it seems that the build process must be run with the `NODE_ENV` environment variable set to `production`. You could use the [`cross-env`](https://www.npmjs.com/package/cross-env) package to set environment variables in NPM scripts so that the build process always has that set. I haven't tested it because I don't have a static file server on my computer, but I think that should work.

**Comment by NoahBres** [comment]
Ah. I did not see the errors in the Application tab. Thank you so much for pointing that out!
I ignored the manifest icon error as I did not think it would cause any errors (favicons are ignored all the time). Apparently not!

I'll look into fixing the manifest!
The build process isn't an issue. Netlify (upon which LRR is hosted) automatically sets the environment variable.
Actually by seeing the error in the Application tab you can surmise that this isn't an issue.

**Comment by NoahBres** [comment]
I don't want to merge anything until it's tested but if you have (or I can) then I am happy to merge it! Or you can make a PR and I'll just make push commits to your PR if it doesn't work. Whatever works!

**Comment by BSFishy** [comment]
I created a PR (#21) with my changes. It looks like Netlify creates a preview site, so even though I can't test locally, it looks like it should be able to be tested there.

**Comment by BSFishy** [comment]
Checking the deployment, the service worker wasn't generated. It looks like [there was an issue in the build process](https://app.netlify.com/sites/learnroadrunner/deploys/6088126a91920000071fec14#L70). I can test again on my machine. Perhaps a difference between the server and development environments?

**Comment by NoahBres** [comment]
Update since this seems to fix it: b45963e860c8e490f66ff4dea1ccab008d01deaa

the vue pwa plugin was in `devDependencies` in `package.json` which are not installed in production. This was my bad. Wish it errored properly though.
